images.go

```diff
@@ -709,23 +709,31 @@ func (r *imageStore) Create(id string, names []string, layer, metadata string, c
 			return nil, fmt.Errorf("image name %q is already associated with image %q: %w", name, image.ID, ErrDuplicateName)
 		}
 	}
-	if created.IsZero() {
-		created = time.Now().UTC()
+	if options.CreationDate.IsZero() {
+		options.CreationDate = time.Now().UTC()
 	}
 
 	image = &Image{
 		ID:             id,
-		Digest:         searchableDigest,
-		Digests:        nil,
+		Digest:         options.Digest,
+		Digests:        copyDigestSlice(options.Digests),
 		Names:          names,
 		TopLayer:       layer,
-		Metadata:       metadata,
+		Metadata:       options.Metadata,
 		BigDataNames:   []string{},
 		BigDataSizes:   make(map[string]int64),
 		BigDataDigests: make(map[string]digest.Digest),
-		Created:        created,
+		Created:        options.CreationDate,
 		Flags:          make(map[string]interface{}),
 	}
+	for _, item := range options.BigData {
+		if item.Digest == "" {
+			item.Digest = digest.Canonical.FromBytes(item.Data)
+		}
+		if err = r.setBigData(image, item.Key, item.Data, item.Digest); err != nil {
+			return nil, err
```

Comment:
It would, but I figured teaching the new `GarbageCollect()` method to handle it in #1484 was simpler.  I left the "incomplete" flag in there?  It's a relic of a design that predates `GarbageCollect()`, so I'll go take it out.

Reply:
No worries, I've been playing a bit of catch-up myself.  Anyway, for some reason I was thinking we'd trend toward routinely calling `GarbageCollect()` at startup in places that, unlike cri-o in a cluster, don't delegate that work to a different process.  That may have been optimistic.  I guess I need to add an "incomplete" state to images.

Reply:
I feel fairly strongly that it is necessary to add deleting data on error paths if this function fails but does not crash.

WRT deleting data on crashes, I’m notably less certain… I can imagine that being a manual process that we don’t automate in principle, and it _might_ be good enough for all I know… especially because automatic incomplete cleanup adds quite a bit of complexity to `startReading`. OTOH images have deterministic IDs, so leaving around data that is _fairly_ likely to conflict with future image creation is risky.

---

One thing to possibly consider is to not add the `BigData` option to container creation, if there there are no current known users and it’s being added only for consistency. That would avoid the need to add a `maxLayerStoreCleanupIterations`-like loop.

Or…, alternatively, making the logic in the three stores _100%_ the same (and perhaps even factoring it out) is a thing to consider. But that would increase the scope quite a bit.

Reply:
Oh, this actually looks simpler than I had thought: if we fail to save those data items, removing the directory where we would have stored the big data items is most of the cleanup we need to do, and everything else is in-memory changes that can be undone pretty cleanly.